flow/operator: Implement clustering for k8s crds (#3824)
grafana/agent: docs/sources/flow/reference/components/prometheus.operator.podmonitors.md

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`enabled` | `bool` | Enables sharing targets with other cluster nodes. | `false` | yes

When the agent is running in [clustered mode][], and `enabled` is set to true,
then this component instance opts-in to participating in
the cluster to distribute scrape load between all cluster nodes.

Clustering assumes that all cluster nodes are running with the same
configuration file, and that all
`prometheus.operator.podmonitors` components that have opted-in to using clustering, over
the course of a scrape interval have the same configuration.

All `prometheus.operator.podmonitors` components instances opting in to clustering use target
labels and a consistent hashing algorithm to determine ownership for each of
the targets between the cluster peers. Then, each peer only scrapes the subset
of targets that it is responsible for, so that the scrape load is distributed.
When a node joins or leaves the cluster, every peer recalculates ownership and
continues scraping with the new target set. This performs better than hashmod
sharding where _all_ nodes have to be re-distributed, as only 1/N of the

> targets ownership is transferred, but is eventually consistent (rather than

Reviewer: ```suggestion
target's ownership is transferred, but is eventually consistent (rather than
```

fully consistent like hashmod sharding is).

> If the agent is _not_ running in clustered mode, then the block is a no-op and

Reviewer: ```suggestion
If the agent is _not_ running in clustered mode, then the block is a no-op, and
```

    }
}
```

> This example will apply additional relabel rules to discovered targets for hashmod sharding. Here we are have 5 shards, and an environment variable is set up to give our current agent's shard.

Reviewer: ```suggestion
This example will apply additional relabel rules to discovered targets for hashmod sharding. Here we have five shards and an environment variable providing our current agent's shard.
```

I'm not sure about the wording here. Is the env var (is that this part? `env("HASHMOD_SHARD")`) storing the name of the agent's shard?  or something else? Depending on what's happening here the wording can change slightly.

Author: Yeah, the idea is that you are running 5 agents, and each has a different value for HASHMOD_SHARD given.
